lib/ABI/FunctionType-revng-c.cpp

```diff
@@ -0,0 +1,424 @@
+  // successful. All the `TypeDependencies` can now be recorded into the
+  // model.
+  M.commit();
+
+  // With all the dependencies out of the way, new `CABIFunctionType` can now
+  // be created,
+  auto [NewType, NewTypePath] = Binary->makeType<model::CABIFunctionType>();
+  NewType.ID() = Function.ID();
+  NewType.CustomName() = Function.CustomName();
+  NewType.OriginalName() = Function.OriginalName();
+  NewType.ABI() = ABI.ABI();
+
+  // And filled in with the argument information.
+  using model::Argument;
+  for (auto &Argument : llvm::concat<Argument>(*RegArguments, *Stack))
+    NewType.Arguments().insert(std::move(Argument));
+  NewType.ReturnType() = *ReturnType;
+
+  revng_log(Log, "Conversion successful.");
+
+  // To finish up the conversion, remove all the references to the old type by
+  // carefully replacing them with references to the new one.
+  return replaceAllUsesWith(Function.key(), NewTypePath, Binary);
+}
+
+using TDManager = TypeDependencyManager;
+template<typename RT>
+std::optional<llvm::SmallVector<model::Argument, 8>>
+TDManager::tryConvertingRegisterArguments(const SortedVector<RT> &Used) {
+  llvm::SmallVector<model::Register::Values, 8> ArgumentRegisters;
+
+  // Rely onto the register state deduction to make sure no "holes" are
+  // present in-between the argument registers.
+  abi::RegisterState::Map Map(model::ABI::getArchitecture(ABI.ABI()));
+  for (const RT &Reg : Used)
+    Map[Reg.Location()].IsUsedForPassingArguments = abi::RegisterState::Yes;
+  auto Enforced = abi::enforceRegisterStateDeductions(Map, ABI);
+  for (auto [Register, Pair] : Enforced)
+    if (abi::RegisterState::shouldEmit(Pair.IsUsedForPassingArguments))
+      ArgumentRegisters.emplace_back(Register);
+
+  // But just knowing which registers we need is not sufficient, we also have
+  // to order them properly.
+  auto Ordered = abi::orderArguments(ArgumentRegisters, ABI);
+
+  llvm::SmallVector<model::Argument, 8> Result;
+  for (model::Register::Values Register : Ordered) {
+    auto CurrentRegisterIterator = Used.find(Register);
+    if (CurrentRegisterIterator != Used.end()) {
+      // If the current register is confirmed to be in use, convert it into
+      // an argument while preserving its type and name.
+      auto &Current = Result.emplace_back();
+      if (Used.at(Register).Type().UnqualifiedType().get() != nullptr)
+        Current.Type() = Used.at(Register).Type();
+      else
+        Current.Type() = { defaultRegisterType(Register), {} };
+      Current.CustomName() = Used.at(Register).CustomName();
+    } else {
+      // This register is unused but we still have to create an argument
+      // for it. Otherwise the resulting function will not be compatible with
+      // the current one: if the "hole" is not preserved, there will be no way
+      // to recreate it at any point in the future, when, for example,
+      // converting back to the original-like representation.
```

Comment:
Yep, not being able to go back is a side effect of effectively producing a `CABIFunctionType` that would be effectively broken. State this clearly.

Reply:
I felt like `will not be compatible` is strong enough wording but sure, there can never be too noticeable of a warning for something this important.

Reply:
Wait, we have two threads about this comment 😅.

I'll close this one in favour of the other one: https://github.com/revng/revng/pull/232#discussion_r1003085151.